### PR TITLE
Add flexibility to ArcGIS Feature Services connector

### PR DESCRIPTION
**Pull request by fvankrieken**
Part of #1004, and touches a bit on #923 

Thought updating ArcGIS Feature Service datasets for GFT was going to be simple. Had a bit of a wrench thrown in when [multiple attempts failed](https://github.com/NYCPlanning/data-engineering/actions/runs/10308316965)

I broke out the commits so that the first commit has a lot of the test data (big part of the added lines in this PR), so commit-by-commit will be easier.

## Context

As a refresher for terms
- **Feature Service** is what this general platform is called
- a **Feature Server** is essentially a dataset collection (a data product if you will), which generally supports a map view
- a **Feature Server Layer** is a classic ESRI layer - one dataset

So our library datasets refer to a single layer of a given Feature Server - in some cases, we have a couple datasets per feature server. We identified these in our templates with a `layer_id` (or assumed it to be `0`.) Turns out layer ids aren't necessarily constant for a given feature server. I think ideally they are supposed to be, but for a handful of layers, the maintainers seemed to delete the previous layer and simply create a new one (with a new id).

Given that, I thought it made sense to
1. not specify id when possible, and grab the only layer there (and throw an error if there's an unexpected layer on the server). Seems to be cleanest for these Feature Servers that are tightly coupled to a single dataset
2. specify name otherwise, and attempt to look up layer by name.

There's a strong argument that generally, it's safer to specify id rather than name... but I don't know if that's something we can rely on here. I would hope so, but it's hard to tell.

## Summary of changes

Separated out logic of models a little bit. Before, we just had a not quite accurately named `FeatureServer` object, shared across library and the connector. I separated things out so we have
- FeatureServer (not just a layer)
- FeatureServerLayer (has name and id, and could contain more metadata if we want)
- FeatureServerLayerDefinition (library-specific - optional name and id)

Separating these out made the different utilities a lot cleaner - not having to worry about if the layer object had its id and name resolved yet.

Beyond that, added and expanded FeatureService utils and added tests, which hadn't existed before

## Test run

[Passing run of calling library for all datasets](https://github.com/NYCPlanning/data-engineering/actions/runs/10325253970)

### CODE REVIEW

**Comment by fvankrieken** [comment]
Just rebased to fix clunky language in a docstring, no code changes since the linked action running all of these

File: dcpy/library/templates/dcp_waterfront_access_map_pow.yml
Comment (fvankrieken) [inline dcpy/library/templates/dcp_waterfront_access_map_pow.yml:1]: These I looked at the feature server linked in the info section to make sure the changes being made were accurate

**Comment by fvankrieken** [comment]
Build is failing due to duplicate rows in a source dataset for templatedb (failing a dbt check). This is still ready to go!

File: dcpy/connectors/esri/arcgis_feature_service.py [hunk 1]
```diff
@@ -12,41 +12,142 @@
...
+            name=feature_server.name,
+            layer_name=layer_name,
+            layer_id=layer_id,
+        )
+        if layer not in layers:
+            raise LookupError(
+                f"Layer '{layer}' not found in feature server {feature_server}"
+            )
+        return layer
+    layers = get_feature_server_layers(feature_server)
```
Comment (alexrichey) [inline dcpy/connectors/esri/arcgis_feature_service.py:72]: ✂️?

Comment (fvankrieken) [reply dcpy/connectors/esri/arcgis_feature_service.py:72]: Oops. Yup!

File: dcpy/connectors/esri/arcgis_feature_service.py [hunk 2]
```diff
@@ -12,41 +12,142 @@
...
+            server=feature_server.server,
+            name=feature_server.name,
+            layer_name=l["name"],
+            layer_id=l["id"],
+        )
+        for l in resp["layers"]
+    ]
+
+
+def resolve_layer(
```
Comment (fvankrieken) [inline dcpy/connectors/esri/arcgis_feature_service.py:46]: I refactored this just slightly - seemed to be a bit cleaner with a `match ... case`. See latest commit

**Review by alexrichey — APPROVED** [review]
LGTM!

**Review by damonmcc — APPROVED** [review]

**Comment by fvankrieken** [comment]
Annoyed at this

https://github.com/python/mypy/issues/12364

(see final comments especially)

**Comment by damonmcc** [comment]
@fvankrieken got some builds failing and seems related to this

GFT and FacDB have the same [error](https://github.com/NYCPlanning/data-engineering/actions/runs/10354811836/job/28661025717?pr=1059#step:6:422) while planning the build:
```bash
ValidationError: 1 validation error for Config
dataset.source.arcgis_feature_server.layer
  Extra inputs are not permitted [type=extra_forbidden, input_value=0, 
input_type=int]
```

**Comment by damonmcc** [comment]
(note from in-person chat) [this action run](https://github.com/NYCPlanning/data-engineering/actions/runs/10355168669) to archive all ArcGIS Feature Server datasets may fix it

**Pull request merged**